scripts/grounding_cell_designer.py

```diff
@@ -0,0 +1,174 @@
+    # Others: isHomologTo
+    bqbio_tags = ['isDescribedBy', 'isEncodedBy', 'is', 'encodes',
+                  'occursIn']
+
+    groundings = defaultdict(list)
+    for tag in bqbio_tags:
+        grounding_elements = \
+            species.findall('sbml:annotation/rdf:RDF/rdf:Description/'
+                            'bqbiol:%s/rdf:Bag/rdf:li' % tag,
+                            namespaces=namespaces)
+        for element in grounding_elements:
+            urn = element.attrib['{http://www.w3.org/1999/02/22-rdf-syntax-ns#}'
+                                 'resource']
+            match = re.match(r'urn:miriam:([^:]+):(.+)', urn)
+            if not match:
+                print('Unmatched urn: %s' % urn)
+                continue
+            else:
+                db_ns, db_id = match.groups()
+                groundings[db_ns].append(db_id)
+    return groundings
+
+
+def add_grounding_element(species, entity_class, db_ns, db_id):
+    # For genes, if we're grounding a protein, we make the encoding aspect
+    # explicit
+    if entity_class == 'PROTEIN' and db_ns == 'HGNC':
+        bqbiol_tag = 'bqbiol:isEncodedBy'
+    # In case a protein is grounded to CHEBI, it's typically a problem, we
+    # skip these
+    elif entity_class == 'PROTEIN' and db_ns == 'CHEBI':
+        return
+    else:
+        bqbiol_tag = 'bqbiol:is'
+
+    tag_sequence = [
+        'sbml:annotation',
+        'rdf:RDF',
+        'rdf:Description',
+        bqbiol_tag,
+        'rdf:Bag',
+    ]
+
+    identifiers_ns = identifiers.get_identifiers_ns(db_ns)
+    grounding_str = 'urn:miriam:%s:%s' % (identifiers_ns, db_id)
+
+    root = species
+    for tag in tag_sequence:
+        element = root.find(tag, namespaces=namespaces)
+        if element is not None:
+            root = element
+        else:
+            new_element = ET.Element(tag)
+            new_element.text = '\n'
+            new_element.tail = '\n'
+            root.append(new_element)
+            root = new_element
+
+    li = ET.Element('rdf:li', attrib={'rdf:resource': grounding_str})
+    root.append(li)
+    return species
+
+
+def dump(et, fname):
+    xml_str = ET.tostring(et.getroot(), xml_declaration=False,
+                          encoding='UTF-8')
+    xml_str = b'<?xml version="1.0" encoding="UTF-8"?>\n' + xml_str
+    xml_str = xml_str.replace(b'ns0:', b'')
+    xml_str = xml_str.replace(b'ns1:', b'')
+    xml_str = xml_str.replace(b' />', b'/>')
+    xml_str = xml_str.replace(b'<html>',
+                              b'<html xmlns="http://www.w3.org/1999/xhtml">')
+    xml_str = xml_str.replace(b'<rdf:RDF>', rdf_str)
+    xml_str = xml_str.replace(b'xmlns:ns0="http://www.sbml.org/sbml/level2/version4"',
+                              b'')
+    xml_str = xml_str.replace(b'xmlns:ns1="http://www.w3.org/1999/xhtml"', b'')
+    with open(fname, 'wb') as fh:
+        fh.write(xml_str)
+
+
+if __name__ == '__main__':
+    base_path = sys.argv[1]
```

Comment:
maybe put this in a click CLI so we remember how to use it later?

Reply:
See 358f6db

Reply:
Thanks!